vmms/dockerSSH.py

```diff
@@ -0,0 +1,323 @@
+            exit(1)
+
+    def boot2dockerVM(self):
+        """ boot2dockerVM - Initializes and starts a boot2docker 
+            VM and sets its environment variables. If boot2docker
+            VM has already been set up on the machine, these steps
+            will simply exit gracefully.
+        """
+        init_ret = -1
+        start_ret = -1
+        env_ret = -1
+        image_ret = -1
+
+        self.log.debug("Initializing boot2docker VM.")
+        init_ret = timeout(['boot2docker', 'init'], 
+                            config.Config.BOOT2DOCKER_INIT_TIMEOUT)
+        
+        self.log.debug("Starting boot2docker VM.")
+        if init_ret == 0:
+            start_ret = timeout(['boot2docker', 'start'], 
+                        config.Config.BOOT2DOCKER_START_TIMEOUT)
+        
+        self.log.debug("Setting environment variables for boot2docker VM.")
+        if start_ret == 0:
+            env_ret = timeout(['boot2docker', 'shellinit'],
+                        config.Config.BOOT2DOCKER_ENV_TIMEOUT)
+        
+        self.log.debug("Pulling the autolab docker image from docker hub.")
+        if env_ret == 0:
+            image_ret = timeout(['docker', 'build', '-t', 
+                        self.config.Config.DOCKER_IMAGE, '.'],
+                        config.Config.DOCKER_IMAGE_BUILD_TIMEOUT)
+
+        if init_ret != 0:
+            raise Exception('Could not initialize boot2docker.')
+        if start_ret != 0:
+            raise Exception('Could not start boot2docker VM.')
+        if env_ret != 0:
+            raise Exception('Could not set environment variables \
+                of boot2docker VM.')
+        if image_ret != 0:
+            raise Exception('Could not pull autolab docker image \
+                from docker hub.')
+
+    def instanceName(self, id, name):
+        """ instanceName - Constructs a VM instance name. Always use
+        this function when you need a VM instance name. Never generate
+        instance names manually.
+        """
+        return "%s-%d-%s" % (config.Config.PREFIX, id, name)
+
+    def domainName(self, vm):
+        """ Returns the domain name that is stored in the vm
+        instance.
+        """
+        return vm.domain_name
+
+    #
+    # VMMS API functions
+    #
+    def initializeVM(self, vm):
+        """ initializeVM -  Start dockerized autograding container by 
+        running a trivially long-running process so that the container
+        continues to run. Otherwise, the container will stop running 
+        once the program has come to completion.
+        """
+        instanceName = self.instanceName(vm.id, vm.name)
+        args = ['docker', 'run', '-d']
+        args.append('--name')
+        args.append(instanceName)
+        args.append(config.Config.DOCKER_IMAGE)
+        args.append('/bin/bash')
+        args.append('-c')
+        args.append('while true; do sleep 1; done')
+        ret = timeout(args, config.Config.INITIALIZEVM_TIMEOUT)
```

Comment:
So do we create the container from the Dockerfile everytime we need a new VM? I feel like it would take way too long to install all the dev binary etc and setup that Docker. Isn't it possible to create a base docker, push it to registry and build from that image in the registry?